### PR TITLE
Allow workflow CWLs to be used as workflow steps

### LINKED ISSUE #2991: CWL Feature: Subworkflows
Link to spec:
- http://www.commonwl.org/v1.0/Workflow.html#SubworkflowFeatureRequirement

Example conformance test(s):

- <span>#</span>40 [v1.0/count-lines8-wf.cwl](https://github.com/common-workflow-language/common-workflow-language/blob/master/v1.0/v1.0/count-lines8-wf.cwl)
- <span>#</span>91 [v1.0/count-lines10-wf.cwl](https://github.com/common-workflow-language/common-workflow-language/blob/master/v1.0/v1.0/count-lines10-wf.cwl)

A/C: This or similar CWL should run to `WorkflowSucceeded` via centaur
Bonus: CWL conformance test passes (may also be a future ticket)

### PR DESCRIPTION
**Pull request by cjllanwarne**
Fixes #2991

### CODE REVIEW

File: cwl/src/main/scala/cwl/FullyQualifiedName.scala [hunk 1]
```diff
@@ -37,18 +37,19 @@ object FileStepAndId {
...
-          case step :: uuid :: id :: Nil => FileStepUUID(file, id, uuid, step)
-          case step :: id :: Nil => FileStepAndId(file, step, id)
-          case id :: Nil => FileAndId(file, id)
-          case _ => throw new RuntimeException(s"malformed FQN: $in")
-        }
-       case _ => throw new RuntimeException(s"malformed FQN: $in")
-     }
+  def apply(in: String): FullyQualifiedName = maybeApply(in).getOrElse(throw new Exception(s"malformed FQN: $in"))
+
+  def maybeApply(in: String): Option[FullyQualifiedName] = {
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/FullyQualifiedName.scala]: um

File: cwl/src/main/scala/cwl/Workflow.scala [hunk 2]
```diff
@@ -92,15 +92,19 @@ case class Workflow private(
...
+              case None => op.name == outputId.id
+            }
+
             for {
               set <- graphFromSteps
               call <- set.collectFirst { case callNode: CallNode if callNode.localName == outputId.stepId => callNode }.
                 toRight(NonEmptyList.one(s"Call Node by name ${outputId.stepId} was not found in set $set"))
-              output <- call.outputPorts.find(_.name == outputId.id).
-                          toRight(NonEmptyList.one(s"looking for ${outputId.id} in call $call output ports ${call.outputPorts}"))
+              output <- call.outputPorts.find(isRightOutputPort).toRight(NonEmptyList.one(s"looking for ${outputId.id} in call $call output ports ${call.outputPorts}"))
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/Workflow.scala]: `toChecked` sugar

File: cwl/src/main/scala/cwl/CwlCodecs.scala [hunk 3]
```diff
@@ -12,23 +12,32 @@ import shapeless.Coproduct
...
 object CwlCodecs {
   import Implicits._
 
   def decodeCwl(cwlWorkflow: String): Either[NonEmptyList[String], Cwl] = {
     //try to parse both and combine errors if they fail
-    (decode[Workflow](cwlWorkflow), decode[CommandLineTool](cwlWorkflow)) match {
-      case (Right(wf), _) => Coproduct[Cwl](wf).asRight
-      case (_, Right(clt)) => Coproduct[Cwl](clt).asRight
-      case (Left(wfError), Left(cltError)) =>
+    (decode[Workflow](cwlWorkflow), decode[CommandLineTool](cwlWorkflow), decode[ExpressionTool](cwlWorkflow)) match {
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/CwlCodecs.scala]: how so?

File: cwl/src/main/scala/cwl/FullyQualifiedName.scala (hunk 1 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/FullyQualifiedName.scala]: Suggestions welcome on a postcard

File: cwl/src/main/scala/cwl/CwlCodecs.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CwlCodecs.scala]: I wanted to `lazy val` the second `decode` back in Dan's PR but it didn't work the way I had hoped so a strict evaluation was necessary, and now an `ExpressionTool` decode has been added. Don't worry about this for your PR, I might look at it over the weekend.

File: cwl/src/main/scala/cwl/package.scala [hunk 4]
```diff
@@ -27,11 +27,12 @@ import wom.types._
  */
 package object cwl extends TypeAliases {
 
-  type Cwl = Workflow :+: CommandLineTool :+: CNil
+  type Cwl = Workflow :+: CommandLineTool :+: ExpressionTool :+: CNil
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: Run can still be a `String`

[the remark this answers is not in the document]
Comment (mcovarr) [reply cwl/src/main/scala/cwl/package.scala]: 😮 what does that even mean?

[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: In CWL, the `Run` is usually a reference to another file. I'm not sure in our "all-in-one CWL"s that's ever true

[the remark this answers is not in the document]
Comment (mcovarr) [reply cwl/src/main/scala/cwl/package.scala]: Hmm I wonder if this is something we should be concerned about supporting...

[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: we do, the new `three-step-subworkflow.cwl` uses a file-based run

[the remark this answers is not in the document]
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: Maybe I should have used this example in the first place... https://github.com/broadinstitute/cromwell/pull/3026/files#diff-3bde2995a68be0859db8b1ea8c89fb7dR20

File: cwl/src/main/scala/cwl/Workflow.scala [hunk 5]
```diff
@@ -92,15 +93,19 @@ case class Workflow private(
...
+              case None => op.name == outputId.id
+            }
+
             for {
               set <- graphFromSteps
               call <- set.collectFirst { case callNode: CallNode if callNode.localName == outputId.stepId => callNode }.
                 toRight(NonEmptyList.one(s"Call Node by name ${outputId.stepId} was not found in set $set"))
-              output <- call.outputPorts.find(_.name == outputId.id).
-                          toRight(NonEmptyList.one(s"looking for ${outputId.id} in call $call output ports ${call.outputPorts}"))
+              output <- call.outputPorts.find(isRightOutputPort).toChecked(s"looking for ${outputId.id} in call $call output ports ${call.outputPorts}")
```
Comment (danbills) [inline cwl/src/main/scala/cwl/Workflow.scala]: ToL consider adding line breaks.

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala [hunk 6]
```diff
@@ -37,7 +37,9 @@ final case class WorkflowStoreSubmitActor(store: WorkflowStore, serviceRegistryA
 
       futureId onComplete {
         case Success(id) =>
-          log.info("Workflow {} submitted.", id)
+          val wfType = cmd.source.workflowType.getOrElse("Unspecified type")
+          val wfTypeVersion = cmd.source.workflowTypeVersion.getOrElse("Unspecified version")
+          log.info("{} ({}) workflow {} submitted", wfType, wfTypeVersion, id)
```
Comment (danbills) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: ToL why not interpolate?

Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala]: this is more efficient as it doesn't interpolate if this is not logged

File: cwl/src/main/scala/cwl/WorkflowStep.scala [hunk 7]
```diff
@@ -60,14 +60,20 @@ case class WorkflowStep(
                      workflowInputs: Map[String, GraphNodeOutputPort]): Checked[Set[GraphNode]] = {
 
     // To avoid duplicating nodes, return immediately if we've already covered this node
-    val haveWeSeenThisStep: Boolean = knownNodes.collect { case TaskCallNode(identifier, _, _, _) => identifier }.contains(unqualifiedStepId)
+    val haveWeSeenThisStep: Boolean = knownNodes.collect {
+      case TaskCallNode(identifier, _, _, _) => identifier
+      case WorkflowCallNode(identifier, _, _, _) => identifier
+      // TODO CWL: Catch known ExpressionTools too
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/WorkflowStep.scala]: good that there's at least a TODO now, but I like what you did before with exploding even better

**Pull request merged**